UOP1_Project/Assets/Scripts/Sound/SoundManager.cs

```diff
@@ -0,0 +1,195 @@
+	private static float MixerValueNormalized(float value)
+	{
+		return  (-(value - 80) / 80) - 1;
+	}
+	private static float NormalizedToMixerValue(float normalizedValue)
+	{
+		return -80 + (normalizedValue * 80);
+	}
+	#endregion
+	#endregion
+
+	#region play sounds functions
+	// Vector3.zero is used as a marker of a 2D sound
+	public SoundEmitter Play2DSound(AudioClip clip, AudioMixerGroup mixer, bool loop = false, float volume = 1, float pitch = 1)
+	{
+		if (!mixer)
+			mixer = masterMixer;
+
+		SoundEmitter foundSoundEmit = GetSoundEmitter();
+
+		foundSoundEmit.PlaySound(clip, mixer, Vector3.zero, loop, volume, pitch);
+
+		return foundSoundEmit;
+	}
+
+	public SoundEmitter PlaySpatialSound(AudioClip clip, AudioMixerGroup mixer, Vector3 position, bool loop = false, float volume = 1, float pitch = 1)
+	{
+		if (!mixer)
+			mixer = masterMixer;
+
+		SoundEmitter foundSoundEmit = GetSoundEmitter();
+
+		foundSoundEmit.PlaySound(clip, mixer, position, loop, volume, pitch);
+
+		return foundSoundEmit;
+	}
+
+	public void StopAllSounds()
+	{
+		foreach (SoundEmitter emit in soundEmitterPool)
+		{
+			if (emit.IsInUse())
+   				emit.StopSound();
+		}
+	}
+	public void StopAllNonLoopSounds()
+	{
+		foreach (SoundEmitter emit in soundEmitterPool)
+		{
+			if (emit.IsInUse() && !emit.IsLooping())
+				emit.StopSound();
+		}
+	}
+	public void StopAllLoopSounds()
+	{
+		foreach (SoundEmitter emit in soundEmitterPool)
+		{
+			if (emit.IsInUse() && emit.IsLooping())
+				emit.StopSound();
+		}
+	}
+	#endregion
+
+	#region SoundEmitters Pool Functions
+	private void InitEmitterPool()
+	{
+		soundEmittersContainer = new GameObject();
+		soundEmittersContainer.name = "Sound Emitters Pool";
+		DontDestroyOnLoad(soundEmittersContainer);
+
+		for (int i = 0; i < initialPoolSize; i++)
+		{
+			CreateNewSoundEmitter();
+		}
+	}
+
+	private SoundEmitter CreateNewSoundEmitter()
+	{
+		GameObject temp = new GameObject();
+		temp.AddComponent<SoundEmitter>();
+		temp.name = "SoundEmitter #" + soundEmitterPool.Count;
+
+		temp.transform.SetParent(soundEmittersContainer.transform);
+
+		soundEmitterPool.Add(temp.GetComponent<SoundEmitter>());
+		return temp.GetComponent<SoundEmitter>();
+	}
+
+	// GetSoundEmitter with return the next SoundEmitter that is not playing any sounds, if all the sound emitters are busy it will extend the pool
+	private SoundEmitter GetSoundEmitter()
+	{
+		SoundEmitter toReturn = null;
+		int startIndex = poolIndex;
+		poolIndex++;
+
+		// Cycles through the list of emitters until it finds one in use or has completed a full cycle
+		while (toReturn == null)
+		{
+			if (poolIndex >= soundEmitterPool.Count)
+				poolIndex = 0;
+
+			if (poolIndex == startIndex) // If none of the sound emitters are available, create a new sound emitter and assign the index to the newly created one
+			{
+				toReturn = CreateNewSoundEmitter();
```

Comment:
I am aware of that and its something mentioned in the initial comment, its something that I'm planning on adding soon.

> This functionality can be expanded to remove SoundEmitter instances past the initial pool once they are no longer needed.

Reply:
Agree that eventually the pool should trim audio sources that haven't been played in a while. Along these lines - should we cap pool at `AudioConfiguration.numVirtualVoices`?

Reply:
I think using AudioConfiguration.numVirtualVoices as maximum would probably defeat the purpose of having a garbage collection since all sounds past this would not work at all.
My initial thought was to garbage collect all the extra instances created past the initial batch.

Reply:
@davejrodriguez I do not think that would be smart. We can have more sounds queued up than would actually be playing. So we can queue up 100 sounds, but only 32 of those are playing at one time for example. If one sound currently playing is done, freeing one of the virtual voices, another sound could chirp in. I am not quite sure if this works like that though, I do not know much about the unity sound system.

If sounds would not start playing either way after being "queued up" like i explained, than your proposal would be the correct choice.

Reply:
>We could use this as a maximum of created instances though, and if that limit is reached just not play this sound.
Not sure what you thought I meant, but maximum created instances is it. And I'm not saying do this instead of garbage collection. I'm saying do it in addition. No point in instantiating 1000 voices if your virtual max is 512.

@Deragaun There is no queuing happening in this code. If 1000 sound emitters start playing, (with current Unity defaults) 32 of them will be playing and you will hear them. 480 will be playing but you will not hear them. 512 will not be playing and you will not hear them. That last 512 is pointless to instantiate. If you're going to add queuing functionality, don't instantiate the sound emitters until you need them and can actually use them.

Reply:
Sorry I misunderstood you then, we agree on that way of doing it then.

Reply:
As previously stated, I don't think it's absolutely crucial for a first PR, especially so early on in the project. Just some ideas to think about.

Reply:
Afaik there is 2 ways of doing this next part of the pooler.
Either we make it so extra instances remove and destroy themselves from the pool after clip.length seconds have passed. or SoundManager removes periodically all instances not in use that that are tagged as instances.

I would probably go with the second one since I think SoundManager should handle everything related to the list to avoid issues, and its also more performant overall given you don't have all the instances updating timer or running coroutines.

Reply:
I was thinking about this too and I agree. I do not think the SoundEmitters should be able to remove themselves from the list. Not their job.

> SoundManager removes periodically all instances not in use that that are tagged as instances.

So, you'll probably want to give it some lifetime where it could exist, inactive, ready to play another clip if one were requested. Otherwise, you lose the performance gains you get by not instantiating/destroying. Could SoundEmitter maintain a timestamp of when it stops being in use? Then every trimming cycle SoundManager could check the current time against the time the emitter stopped and destroy the ones greater than some lifetime? This all could be done in SoundManager too (marking first time !IsInUse and destroying if detected again longer than lifetime), but _could_ generate false positives depending on trim frequency.

Reply:
Sounds good, I'll have a first pass implementation tonight.